### PR TITLE
feat(YouTube - Download Playlist Button): Add playlist download button

### PR DESCRIPTION
**Pull request by Francesco146**
ref: https://github.com/anddea/revanced-patches/pull/761

### CODE REVIEW

File: app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java [hunk 1]
```diff
@@ -104,12 +104,20 @@ public static boolean inAppDownloadButtonOnClick(String videoId) {
     }
 
     public static void launchExternalDownloader() {
-        launchExternalDownloader(VideoInformation.getVideoId());
+        launchExternalDownloader(VideoInformation.getVideoId(), false);
     }
 
-    public static void launchExternalDownloader(@NonNull String videoId) {
+    public static void launchExternalDownloader(@NonNull String resourceId, boolean isPlaylist) {
```
Comment (fromhgbwithluv) [inline app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java]: We can seperate this function into `launchVideoExternalDownloader` and `launchPlaylistExternalDownloader`

File: app/src/main/java/app/revanced/integrations/youtube/patches/misc/DownloadPlaylistButton.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package app.revanced.integrations.youtube.patches.misc;
+
+import app.revanced.integrations.youtube.settings.Settings;
+import app.revanced.integrations.youtube.utils.VideoUtils;
+
+@SuppressWarnings("unused")
+public class DownloadPlaylistButton {
```
Comment (fromhgbwithluv) [inline app/src/main/java/app/revanced/integrations/youtube/patches/misc/DownloadPlaylistButton.java]: Uhm the name is not correct bruh. Also move the integrations of Video Hook to this class

File: app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java [hunk 3]
```diff
@@ -107,8 +107,9 @@ public static void launchExternalDownloader() {
         launchExternalDownloader(VideoInformation.getVideoId());
     }
 
-    public static void launchExternalDownloader(@NonNull String videoId) {
+    public static void launchExternalDownloader(@NonNull String videoID) {
```
Comment (fromhgbwithluv) [inline app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java]: `videoId` look better, and match all of the parameter in integrations 😭

File: app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java [hunk 4]
```diff
@@ -130,6 +133,27 @@ public static void launchExternalDownloader(@NonNull String videoId) {
         }
     }
 
+    public static void launchPlaylistExternalDownloader(@NonNull String playlistId) {
+        try {
+            // use default downloader (YTDLnis) if user tries to download a playlist
+            // because most downloaders don't support playlist download
+            String downloaderPackageName = externalDownloaderPackageName.defaultValue;
```
Comment (fromhgbwithluv) [inline app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java]: why not hardcoded the value there 👀 So if one day, inotia change the default value to Seal, or new downloader, then this feature will still worked as intended

File: app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java [hunk 5]
```diff
@@ -130,6 +133,27 @@ public static void launchExternalDownloader(@NonNull String videoId) {
...
+
+            final String content = String.format("https://www.youtube.com/playlist?list=%s", playlistId);
+
+            if (!checkPackageIsEnabled()) {
```
Comment (fromhgbwithluv) [inline app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java]: `checkPackageIsEnabled()`: We should add package name as a parameter in this function

Comment (Francesco146) [reply app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java:144]: why?

Comment (fromhgbwithluv) [reply app/src/main/java/app/revanced/integrations/youtube/utils/VideoUtils.java]: > why?

this function check if the setted downloader is installed, not checking YTDLnis

**Pull request merged**